File: app/src/main/java/it/notes/ecosystem/media/MediaBundle.kt
package it.notes.ecosystem.media

import it.notes.ecosystem.domain.*
import it.notes.ecosystem.data.BackupReader
import it.notes.ecosystem.data.BackupWriter
import org.json.JSONObject
import org.json.JSONArray
import java.io.*
import java.nio.file.Files
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

class MediaBundlePreview(val snapshot:BackupSnapshot,val staging:File,val keys:Set<String>,val bytes:Long):Closeable {
    fun install(store:AttachmentFiles) {keys.forEach {key->store.put(key,File(staging,key).inputStream().use {readLimited(it,Attachments.FILE_LIMIT)})}}
    override fun close() {staging.listFiles().orEmpty().forEach {it.delete()};staging.delete()}
}
object MediaBundle {
    fun write(snapshot:BackupSnapshot,store:AttachmentFiles,output:OutputStream,onProgress:(Int,Int)->Unit={_,_->}) {
        val keys=Attachments.keys(snapshot).sorted();require(keys.size<=Attachments.MAX_FILES)
        val json=BackupWriter.json(snapshot).toByteArray(Charsets.UTF_8)
        require(json.size<=5*1024*1024){"Il backup delle note supera 5 MiB."}
        // Validate every referenced file before creating a supposedly complete backup.
        var total=0L
        keys.forEach {key->total+=store.read(key).size;require(total<=Attachments.BATCH_LIMIT){"Gli allegati del backup superano 64 MiB."}}
        ZipOutputStream(output).use {zip ->
            fun entry(name:String,bytes:ByteArray) {zip.putNextEntry(ZipEntry(name));zip.write(bytes);zip.closeEntry()}
            entry("bundle.json",JSONObject().put("format","notes-ecosystem-media").put("version",1).put("assets",JSONArray(keys)).toString().toByteArray())
            entry("backup.json",json)
            keys.forEachIndexed {i,key->entry("assets/$key",store.read(key));onProgress(i+1,keys.size)}
            snapshot.notes.filter {it.deletedAt==null}.forEachIndexed {i,n->
                when (n.sketch?.kind) {
                    VisualDocumentKind.SKETCH -> entry("disegni/${i+1}.sketch.json", n.body.toByteArray())
                    VisualDocumentKind.WHITEBOARD -> entry("lavagne/${i+1}.whiteboard.json", n.body.toByteArray())
                    null -> entry("note/${i+1}.md", ("# "+n.title+"\n\n"+Attachments.portable(n.body)).toByteArray())
                }
            }
            entry("LEGGIMI.txt","Importa questo ZIP completo in Notes 0.17+. backup.json da solo non contiene i file multimediali. assets contiene originali verificati SHA-256. Ricerche salvate, revisioni locali e timer attivo non inclusi. File non cifrati.".toByteArray())
        }
    }
    fun read(input:InputStream,tempRoot:File):MediaBundlePreview {
        check(tempRoot.isDirectory || tempRoot.mkdirs())
        val stage=Files.createTempDirectory(tempRoot.toPath(),"bundle-").toFile()
        var snapshot:BackupSnapshot?=null;var manifest:Set<String>?=null;val assets=mutableSetOf<String>();val seen=mutableSetOf<String>()
        var total=0L;var compressed=0L;var assetBytes=0L
        val bounded=object:FilterInputStream(input) {
            override fun read():Int {val v=super.read();if(v>=0){compressed++;require(compressed<=80L*1024*1024)};return v}
            override fun read(b:ByteArray,off:Int,len:Int):Int {val n=`in`.read(b,off,len);if(n>0){compressed+=n;require(compressed<=80L*1024*1024)};return n}
        }
        try {
            ZipInputStream(bounded).use {zip ->
                while(true) {
                    val e=zip.nextEntry ?: break
                    require(!e.isDirectory && seen.add(e.name) && seen.size<=11005){"Voce ZIP duplicata o non valida."}
                    val key=e.name.removePrefix("assets/")
                    val asset=e.name.startsWith("assets/") && Attachments.validKey(key)
                    val allowed=asset || e.name in listOf("bundle.json","backup.json","LEGGIMI.txt") || e.name.matches(Regex("note/[0-9]+\\.md|disegni/[0-9]+\\.sketch\\.json|lavagne/[0-9]+\\.whiteboard\\.json"))
                    require(allowed){"Percorso ZIP non consentito."}
                    val limit=when {asset->Attachments.FILE_LIMIT;e.name=="backup.json"->5*1024*1024;e.name=="bundle.json"->100000;e.name=="LEGGIMI.txt"->10000;else->Attachments.FILE_LIMIT}
                    val bytes=readLimited(zip,limit);total+=bytes.size
                    require(total<=80L*1024*1024){"Archivio espanso oltre 80 MiB."}
                    when {
                        asset->{require(assets.size<Attachments.MAX_FILES);Attachments.verify(key,bytes);assets+=key;assetBytes+=bytes.size;require(assetBytes<=Attachments.BATCH_LIMIT);File(stage,key).writeBytes(bytes)}
                        e.name=="backup.json"->snapshot=BackupReader.parse(strictText(bytes))
                        e.name=="bundle.json"->{val o=JSONObject(strictText(bytes));require(o.getString("format")=="notes-ecosystem-media" && o.getInt("version")==1);val a=o.getJSONArray("assets");require(a.length()<=Attachments.MAX_FILES);val list=(0 until a.length()).map {a.getString(it)};require(list.all(Attachments::validKey) && list.distinct().size==list.size);manifest=list.toSet()}
                    }
                    zip.closeEntry()
                }
            }
            val data=checkNotNull(snapshot){"backup.json mancante."}
            require(manifest!=null && manifest==assets && assets==Attachments.keys(data)){"Allegati mancanti o non coerenti con le note."}
            return MediaBundlePreview(data,stage,assets,assetBytes)
        } catch(t:Throwable) {stage.listFiles().orEmpty().forEach {it.delete()};stage.delete();throw t}
    }
    private fun strictText(bytes:ByteArray)=Charsets.UTF_8.newDecoder().onMalformedInput(java.nio.charset.CodingErrorAction.REPORT)
        .onUnmappableCharacter(java.nio.charset.CodingErrorAction.REPORT).decode(java.nio.ByteBuffer.wrap(bytes)).toString()
}

File: flutter_app/lib/src/domain/media_bundle.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:archive/archive.dart';

import 'attachments.dart';
import 'backup.dart';
import 'visual_documents.dart';

class MediaBundlePreview {
  const MediaBundlePreview({
    required this.snapshot,
    required this.assets,
    this.properties = const {},
    this.knowledge = const {},
    this.derivatives = const {},
  });

  final BackupSnapshot snapshot;
  final Map<String, Uint8List> assets;
  final Map<String, Object?> properties;
  final Map<String, Object?> knowledge;
  final Map<String, Object?> derivatives;

  int get assetBytes =>
      assets.values.fold<int>(0, (sum, bytes) => sum + bytes.length);
}

abstract final class MediaBundle {
  static const maxArchiveBytes = 80 * 1024 * 1024;
  static const maxAssetBatchBytes = 64 * 1024 * 1024;
  static const format = 'notes-ecosystem-media';
  static const version = 2;
  static const maxSidecarBytes = 8 * 1024 * 1024;

  static Future<Uint8List> encode(
    BackupSnapshot snapshot,
    AttachmentStore store, {
    Map<String, Object?> properties = const {},
    Map<String, Object?> knowledge = const {},
    Map<String, Object?> derivatives = const {},
  }) async {
    final keys = referencedKeys(snapshot).toList()..sort();
    final assets = <String, Uint8List>{};
    for (final key in keys) {
      assets[key] = await store.read(key);
    }
    return encodeLoaded(
      snapshot,
      assets,
      properties: properties,
      knowledge: knowledge,
      derivatives: derivatives,
    );
  }

  static Uint8List encodeLoaded(
    BackupSnapshot snapshot,
    Map<String, Uint8List> assets, {
    Map<String, Object?> properties = const {},
    Map<String, Object?> knowledge = const {},
    Map<String, Object?> derivatives = const {},
  }) {
    BackupCodec.validate(snapshot);
    final keys = referencedKeys(snapshot).toList()..sort();
    if (keys.length > Attachments.maxFiles ||
        assets.length != keys.length ||
        !assets.keys.toSet().containsAll(keys) ||
        !keys.toSet().containsAll(assets.keys)) {
      throw const FormatException(
        'Allegati mancanti o non coerenti con le note.',
      );
    }

    final backup = BackupCodec.encode(snapshot);
    if (utf8.encode(backup).length > BackupCodec.maxBytes) {
      throw const FormatException('Il backup delle note supera 5 MiB.');
    }

    final archive = Archive();
    archive.add(
      ArchiveFile.string(
        'bundle.json',
        jsonEncode({
          'format': format,
          'version': version,
          'assets': keys,
          'sidecars': const [
            'properties.json',
            'knowledge.json',
            'derivatives.json',
          ],
        }),
      ),
    );
    archive.add(ArchiveFile.string('backup.json', backup));
    for (final entry in <String, Map<String, Object?>>{
      'properties.json': properties,
      'knowledge.json': knowledge,
      'derivatives.json': derivatives,
    }.entries) {
      final encoded = jsonEncode(entry.value);
      if (utf8.encode(encoded).length > maxSidecarBytes) {
        throw FormatException('${entry.key} supera 8 MiB.');
      }
      archive.add(ArchiveFile.string(entry.key, encoded));
    }

    var assetBytes = 0;
    for (final key in keys) {
      final bytes = assets[key]!;
      Attachments.verify(key, bytes);
      assetBytes += bytes.length;
      if (assetBytes > maxAssetBatchBytes) {
        throw const FormatException(
          'Gli allegati del backup superano 64 MiB.',
        );
      }
      archive.add(ArchiveFile.bytes('assets/$key', bytes));
    }

    var textIndex = 0;
    var sketchIndex = 0;
    var boardIndex = 0;
    for (final note in snapshot.notes.where((note) => !note.isDeleted)) {
      if (note.sketchJson == null) {
        textIndex++;
        archive.add(
          ArchiveFile.string(
            'note/$textIndex.md',
            '# ${note.title}\n\n${_portableBody(note.body)}',
          ),
        );
      } else {
        final info = VisualInfo.decode(note.sketchJson!);
        if (info.kind == VisualInfoKind.sketch) {
          sketchIndex++;
          archive.add(
            ArchiveFile.string(
              'disegni/$sketchIndex.sketch.json',
              note.body,
            ),
          );
        } else {
          boardIndex++;
          archive.add(
            ArchiveFile.string(
              'lavagne/$boardIndex.whiteboard.json',
              note.body,
            ),
          );
        }
      }
    }

    archive.add(
      ArchiveFile.string(
        'LEGGIMI.txt',
        'Backup completo Notes 0.36+. '
            'backup.json contiene il workspace canonico; assets contiene gli originali '
            'verificati SHA-256; properties/knowledge/derivatives contengono i sidecar '
            'portabili. Ricerche salvate, revisioni locali e timer Focus attivo non '
            'sono inclusi. Archivio non cifrato.',
      ),
    );

    final bytes = ZipEncoder().encodeBytes(archive);
    if (bytes.length > maxArchiveBytes) {
      throw const FormatException('Il backup ZIP supera 80 MiB.');
    }
    return bytes;
  }

  static MediaBundlePreview decode(Uint8List bytes) {
    if (bytes.isEmpty || bytes.length > maxArchiveBytes) {
      throw const FormatException('Archivio ZIP vuoto o oltre 80 MiB.');
    }

    final archive = ZipDecoder().decodeBytes(bytes, verify: true);
    if (archive.length > 11005) {
      throw const FormatException('Troppe voci nel backup.');
    }

    final seen = <String>{};
    String? backupText;
    Set<String>? manifest;
    int? bundleVersion;
    Set<String> declaredSidecars = const {};
    Map<String, Object?> properties = const {};
    Map<String, Object?> knowledge = const {};
    Map<String, Object?> derivatives = const {};
    final assets = <String, Uint8List>{};
    var expandedBytes = 0;
    var assetBytes = 0;

    for (final entry in archive) {
      if (!entry.isFile || !seen.add(entry.name)) {
        throw const FormatException('Voce ZIP duplicata o non valida.');
      }
      final name = entry.name;
      final isAsset = name.startsWith('assets/');
      final assetKey = isAsset ? name.substring('assets/'.length) : '';
      final allowed = isAsset && Attachments.validKey(assetKey) ||
          name == 'bundle.json' ||
          name == 'backup.json' ||
          name == 'properties.json' ||
          name == 'knowledge.json' ||
          name == 'derivatives.json' ||
          name == 'LEGGIMI.txt' ||
          RegExp(r'^note/[0-9]+\.md$').hasMatch(name) ||
          RegExp(r'^disegni/[0-9]+\.sketch\.json$').hasMatch(name) ||
          RegExp(r'^lavagne/[0-9]+\.whiteboard\.json$').hasMatch(name);
      if (!allowed) {
        throw const FormatException('Percorso ZIP non consentito.');
      }

      final data = entry.readBytes();
      if (data == null) {
        throw const FormatException('Voce ZIP non leggibile.');
      }
      expandedBytes += data.length;
      if (expandedBytes > maxArchiveBytes) {
        throw const FormatException('Archivio espanso oltre 80 MiB.');
      }

      if (isAsset) {
        if (assets.length >= Attachments.maxFiles ||
            data.isEmpty ||
            data.length > Attachments.fileLimit) {
          throw const FormatException('Allegato ZIP non valido.');
        }
        final typed = Uint8List.fromList(data);
        Attachments.verify(assetKey, typed);
        assetBytes += typed.length;
        if (assetBytes > maxAssetBatchBytes) {
          throw const FormatException(
            'Gli allegati del backup superano 64 MiB.',
          );
        }
        assets[assetKey] = typed;
      } else if (name == 'backup.json') {
        if (data.length > BackupCodec.maxBytes) {
          throw const FormatException('backup.json supera 5 MiB.');
        }
        backupText = utf8.decode(data, allowMalformed: false);
      } else if (name == 'properties.json' ||
          name == 'knowledge.json' ||
          name == 'derivatives.json') {
        if (data.length > maxSidecarBytes) {
          throw FormatException('$name supera 8 MiB.');
        }
        final decoded = jsonDecode(utf8.decode(data, allowMalformed: false));
        if (decoded is! Map) {
          throw FormatException('$name non valido.');
        }
        final mapped =
            decoded.map((key, value) => MapEntry(key.toString(), value));
        if (name == 'properties.json') properties = mapped;
        if (name == 'knowledge.json') knowledge = mapped;
        if (name == 'derivatives.json') derivatives = mapped;
      } else if (name == 'bundle.json') {
        if (data.length > 100000) {
          throw const FormatException('Manifest backup troppo grande.');
        }
        final root = jsonDecode(utf8.decode(data, allowMalformed: false));
        if (root is! Map ||
            root['format'] != format ||
            root['version'] is! num ||
            (root['version'] as num).toInt() < 1 ||
            (root['version'] as num).toInt() > version ||
            root['assets'] is! List) {
          throw const FormatException('Manifest backup non valido.');
        }
        bundleVersion = (root['version'] as num).toInt();
        if (bundleVersion! >= 2) {
          if (root['sidecars'] is! List) {
            throw const FormatException('Sidecar backup mancanti.');
          }
          declaredSidecars = (root['sidecars'] as List)
              .map((value) => value.toString())
              .toSet();
          const required = {
            'properties.json',
            'knowledge.json',
            'derivatives.json',
          };
          if (!declaredSidecars.containsAll(required) ||
              declaredSidecars.length != required.length) {
            throw const FormatException('Elenco sidecar backup non valido.');
          }
        }
        final list = (root['assets'] as List)
            .map((value) => value.toString())
            .toList(growable: false);
        if (list.length > Attachments.maxFiles ||
            list.toSet().length != list.length ||
            list.any((key) => !Attachments.validKey(key))) {
          throw const FormatException('Elenco allegati non valido.');
        }
        manifest = list.toSet();
      }
    }

    if (backupText == null || manifest == null || bundleVersion == null) {
      throw const FormatException('backup.json o bundle.json mancante.');
    }
    if (bundleVersion! >= 2 &&
        (properties.isEmpty && knowledge.isEmpty && derivatives.isEmpty)) {
      // Empty maps are valid individually, but all three files must have been
      // present as declared. Presence is guaranteed by the ZIP seen-set below.
      const required = {
        'properties.json',
        'knowledge.json',
        'derivatives.json',
      };
      if (!seen.containsAll(required)) {
        throw const FormatException('Sidecar backup dichiarati ma mancanti.');
      }
    }
    final snapshot = BackupCodec.decode(backupText);
    final referenced = referencedKeys(snapshot);
    if (manifest.length != assets.length ||
        !manifest.containsAll(assets.keys) ||
        !assets.keys.toSet().containsAll(manifest) ||
        !referenced.containsAll(manifest) ||
        !manifest.containsAll(referenced)) {
      throw const FormatException(
        'Allegati mancanti o non coerenti con le note.',
      );
    }

    return MediaBundlePreview(
      snapshot: snapshot,
      assets: assets,
      properties: properties,
      knowledge: knowledge,
      derivatives: derivatives,
    );
  }

  static Future<void> installAssets(
    MediaBundlePreview preview,
    AttachmentStore store,
  ) async {
    for (final entry in preview.assets.entries) {
      await store.put(entry.key, entry.value);
    }
  }

  static Set<String> referencedKeys(BackupSnapshot snapshot) {
    final keys = <String>{};
    for (final note in snapshot.notes) {
      if (note.sketchJson == null) {
        keys.addAll(Attachments.refs(note.body).map((ref) => ref.key));
      }
    }
    for (final draft in snapshot.drafts) {
      keys.addAll(Attachments.refs(draft.body).map((ref) => ref.key));
    }
    return keys;
  }

  static String _portableBody(String body) => body.replaceAllMapped(
        RegExp(r'notes-asset://([a-f0-9]{64}\.[a-z0-9]{2,4})'),
        (match) => 'assets/${match.group(1)}',
      );
}
